pkg/reconciler/shared/tektonconfig/pipeline/pipeline.go

```diff
@@ -20,35 +20,51 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"log"
 	"reflect"
 
 	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
 
 	op "github.com/tektoncd/operator/pkg/client/clientset/versioned/typed/operator/v1alpha1"
-	operatorv1alpha1 "github.com/tektoncd/operator/pkg/client/clientset/versioned/typed/operator/v1alpha1"
 	"github.com/tektoncd/operator/pkg/reconciler/common"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
-	"knative.dev/pkg/test/logging"
 )
 
-func CreatePipelineCR(ctx context.Context, instance v1alpha1.TektonComponent, client operatorv1alpha1.OperatorV1alpha1Interface) error {
-	configInstance := instance.(*v1alpha1.TektonConfig)
-	if _, err := ensureTektonPipelineExists(ctx, client.TektonPipelines(), configInstance); err != nil {
-		return errors.New(err.Error())
+func EnsureTektonPipelineExists(ctx context.Context, clients op.TektonPipelineInterface, config *v1alpha1.TektonConfig) (*v1alpha1.TektonPipeline, error) {
+	tpCR, err := GetPipeline(ctx, clients, v1alpha1.PipelineResourceName)
+
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			tpCR = &v1alpha1.TektonPipeline{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: v1alpha1.PipelineResourceName,
+				},
+				Spec: v1alpha1.TektonPipelineSpec{
+					CommonSpec: v1alpha1.CommonSpec{
+						TargetNamespace: config.Spec.TargetNamespace,
+					},
+					Pipeline: config.Spec.Pipeline,
+					Config:   config.Spec.Config,
+				},
+			}
+			_, err = clients.Create(ctx, tpCR, metav1.CreateOptions{})
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, v1alpha1.RECONCILE_AGAIN_ERR
+		}
 	}
-	if _, err := waitForTektonPipelineState(ctx, client.TektonPipelines(), v1alpha1.PipelineResourceName,
-		isTektonPipelineReady); err != nil {
-		log.Println("TektonPipeline is not in ready state: ", err)
-		return err
+
+	ok, err := isTektonPipelineReady(tpCR, err)
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, v1alpha1.RECONCILE_AGAIN_ERR
```

Comment:
in the reconcile logic, we need to check if an upadate is needed before checking isReady. If we check ifReady before checking if an update is needed, then the reconciler will never reach the update check.

so the reconciler has to check the cluster state in the following order.
1. check if exists
2. check if update needed
3. check if ready

Reply:
yeah make sense
I had one question, in both the states, i.e. while updating and while creation we need to check if the pipeline is ready or not right ?

Reply:
while creation
1. create api call then reconcile again if no error
1. in the next loop if update needed make update api call, then reconcile again if no error, if update is not needed go to next check
1. in the next loop if update was needed or in the same loop if update was not needed,  check for ready condition, if not ready reconcile again, if err return err

so the check cascade in different execution context of the reconcile function not in the same running instance (context) of reconciler.